demo/ARContainerViewController.swift

```diff
@@ -10,12 +10,10 @@ final class ARContainerViewController: UIViewController {
 
     private lazy var mapViewController = ARMapNavigationController()
     private lazy var arViewController = VisionARViewController()
-    
-    weak var navigationDelegate: NavigationManagerDelegate? {
-        didSet {
-            arViewController.navigationDelegate = navigationDelegate
-        }
-    }
+
+    weak var navigationDelegate: NavigationManagerDelegate?
+    private var navigationService: NavigationService?
+    private var routeHasChanged: Bool = true
```

Comment:
Not sure how, but probably we should remove `routeHasChanged` state from ViewController if it's expanded as a transition state to call delegate's methods. Also I'd try to use delegate methods as much as possible. Smth like (didn't try to compile):

```swift
    func navigationService(_ service: NavigationService, didUpdate progress: RouteProgress, with location: CLLocation, rawLocation: CLLocation) {
        instructionsLabel.isHidden = false
        instructionsLabel.text = progress.currentLegProgress.currentStep.instructions
    }

    func navigationService(_ service: NavigationService, didArriveAt waypoint: Waypoint) -> Bool {
        navigationDelegate?.navigationServiceArrivedAtDestination()
        return true
    }

    func navigationService(_ service: NavigationService, didRerouteAlong route: MapboxDirections.Route, at location: CLLocation?, proactive: Bool) {
        navigationDelegate?.navigationService(didUpdate: Route(route: route))
    }
```

Reply:
done